gc/base/MemorySubSpace.cpp

```diff
@@ -1127,6 +1127,36 @@ MM_MemorySubSpace::canExpand(MM_EnvironmentBase* env, uintptr_t expandSize)
 	return false;
 }
 
+/**
+ * Compare the specified expand amount with -XsoftMX value
+ * @return Updated expand size
+ */
+uintptr_t
+MM_MemorySubSpace::adjustExpansionWithinSoftMax(MM_EnvironmentBase *env, uintptr_t expandSize, uintptr_t minimumBytesRequired)
+{
+	if (0 != _extensions->softMx) {
+		MM_Heap *heap = _extensions->getHeap();
+		uintptr_t activeMemorySize = heap->getActiveMemorySize();
+		OMRPORT_ACCESS_FROM_OMRPORT(env->getPortLibrary());
+
+		if ((0 != minimumBytesRequired) && ((activeMemorySize + minimumBytesRequired) > _extensions->softMx)) {
+			if (J9_EVENT_IS_HOOKED(env->getExtensions()->omrHookInterface, J9HOOK_MM_OMR_OOM_DUE_TO_SOFTMX)) {
+				ALWAYS_TRIGGER_J9HOOK_MM_OMR_OOM_DUE_TO_SOFTMX(env->getExtensions()->omrHookInterface, env->getOmrVMThread(), omrtime_hires_clock(),
+						heap->getMaximumMemorySize(), heap->getActiveMemorySize(), env->getExtensions()->softMx, minimumBytesRequired);
+			}
+		}
+		
+		if (_extensions->softMx < activeMemorySize) {
+			/* If our softmx is smaller than our currentsize, we should be contracting not expanding */
+			expandSize = 0;
+		} else if ((activeMemorySize + expandSize) > _extensions->softMx) {
+			/* We would go past our -XsoftMx so just expand up to it instead */
+			expandSize = _extensions->softMx - activeMemorySize;
```

Comment:
I know this is existing code, but while touching it, it would be nice to introduce trace points for those 2 cases when adjustment is done.